### PR TITLE
msi-ec-kmods: init at 0-unstable-2024-09-19

### PR DESCRIPTION
**Pull request by m1dugh**
## Adds driver `msi-ec-kmods`

Adds the driver for [msi embedded controller](https://github.com/BeardOverflow/msi-ec).

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin

- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of driver
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by NotAShelf** [comment]
There seems to be several issues with this PR.

1. You will want to format new files with nixfmt-rfc-style. You might have noticed that the formatting PR is failing
2. Editorconfig check is also failing, see if your editor integrates with it and try to conform to nixpkgs' editorconfig
3. Please rebase this PR, a rogue PR has managed to sneak in your git history, you might want to drop it.
4. Adding yourself to [maintainers/maintainer-list.nix](https://github.com/NixOS/nixpkgs/pull/343038/files#diff-c1fcec3afa1da96245de1d7954ca721c979a5f25aee0e8b59d59d3d21e31f8fa) must be done in a separate commit. You will need to rebase again for this.

**Review by NotAShelf — CHANGES_REQUESTED** [review]
There seems to be several issues with this PR.

1. You will want to format new files with nixfmt-rfc-style. You might have noticed that the formatting PR is failing
2. Editorconfig check is also failing, see if your editor integrates with it and try to conform to nixpkgs' editorconfig
3. Please rebase this PR, a rogue PR has managed to sneak in your git history, you might want to drop it.
4. Adding yourself to [maintainers/maintainer-list.nix](https://github.com/NixOS/nixpkgs/pull/343038/files#diff-c1fcec3afa1da96245de1d7954ca721c979a5f25aee0e8b59d59d3d21e31f8fa) must be done in a separate commit. You will need to rebase again for this.

File: pkgs/os-specific/linux/msi-ec/default.nix
Comment (GetPsyched) [inline pkgs/os-specific/linux/msi-ec/default.nix:7]: AFAIK this notation isn't needed here.
```suggestion
```

**Comment by m1dugh** [comment]
I have fixed most issues, I am not sure to understand the rogue PR part.

**Comment by NotAShelf** [comment]
Sorry, I meant a rogue *commit**. Seems to be fixed though, so no worries.

File: pkgs/os-specific/linux/msi-ec/default.nix
Comment (GetPsyched) [inline pkgs/os-specific/linux/msi-ec/default.nix:11]: Correct me if I'm wrong here, but I didn't find any release tag in the repo or the one it's forked on.

Either way, the convention in nixpkgs is to do the following when not using a release tag:
```suggestion
  version = "unstable-2024-09-19";
```

File: pkgs/os-specific/linux/msi-ec/default.nix
Comment (GetPsyched) [inline pkgs/os-specific/linux/msi-ec/default.nix:9]:
```suggestion
stdenv.mkDerivation {
```
https://nix.dev/guides/best-practices#recursive-attribute-set-rec

**Review by GetPsyched — CHANGES_REQUESTED** [review]

File: pkgs/os-specific/linux/msi-ec/default.nix
Comment (NotAShelf) [reply pkgs/os-specific/linux/msi-ec/default.nix:7]: Correct.

File: pkgs/os-specific/linux/msi-ec/default.nix
Comment (NotAShelf) [reply pkgs/os-specific/linux/msi-ec/default.nix:11]: The current pattern is `0-unstable-yyyy-mm-dd`. The addition of `0-` is new, but required.

File: pkgs/os-specific/linux/msi-ec/default.nix
Comment (NotAShelf) [reply pkgs/os-specific/linux/msi-ec/default.nix:9]: This is an unsubstantial nit. Though I also prefer avoiding `rec` (there is finalAttrs too, as a solution) there is no clear guideline on this.

File: pkgs/os-specific/linux/msi-ec/default.nix
Comment (NotAShelf) [inline pkgs/os-specific/linux/msi-ec/default.nix:17]: Is there a reason this is using your fork? From what I can tell, upstream is still maintained and there are no technical changes in your  fork.

File: pkgs/os-specific/linux/msi-ec/default.nix
Comment (NotAShelf) [inline pkgs/os-specific/linux/msi-ec/default.nix:48]:
```suggestion
  meta = {
    description = "Kernel modules for MSI Embedded controller";
    homepage = "https://github.com/m1dugh/msi-ec";
    license = lib.licenses.gpl2;
    maintainers = [ lib.maintainers.m1dugh ];
    platforms = lib.platforms.linux;
    broken = lib.kernel.kernelOlder "6.2";
  };
```

Nested `with lib;` is discouraged. You might want to be more explicit.

**Comment by m1dugh** [comment]
Should be all good now

File: pkgs/os-specific/linux/msi-ec/default.nix
Comment (m1dugh) [reply pkgs/os-specific/linux/msi-ec/default.nix:17]: There is still a PR running on the original repo, I intend to change it as soon as the PR is approved

**Review by NotAShelf — COMMENTED** [review]
Derivation LGTM. As a last remark, you will want to update the commit message and the PR title to both show `msi-ec-kmods: init at 0-unstable-2024-09-19`

**Review by GetPsyched — APPROVED** [review]
Derivation LGTM

**Review by NotAShelf — APPROVED** [review]

**Comment by NotAShelf** [comment]
Result of `nixpkgs-review pr 343038` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>10 packages marked as broken and skipped:</summary>
  <ul>
    <li>linuxKernel.packages.linux_4_19.msi-ec</li>
    <li>linuxKernel.packages.linux_4_19_hardened.msi-ec</li>
    <li>linuxKernel.packages.linux_5_10.msi-ec</li>
    <li>linuxKernel.packages.linux_5_10_hardened.msi-ec</li>
    <li>linuxKernel.packages.linux_5_15.msi-ec</li>
    <li>linuxKernel.packages.linux_5_15_hardened.msi-ec</li>
    <li>linuxKernel.packages.linux_5_4.msi-ec</li>
    <li>linuxKernel.packages.linux_5_4_hardened.msi-ec</li>
    <li>linuxKernel.packages.linux_6_1.msi-ec</li>
    <li>linuxKernel.packages.linux_6_1_hardened.msi-ec</li>
  </ul>
</details>
<details>
  <summary>2 packages failed to build:</summary>
  <ul>
    <li>linuxKernel.packages.linux_6_11.msi-ec</li>
    <li>linuxKernel.packages.linux_latest_libre.msi-ec</li>
  </ul>
</details>
<details>
  <summary>8 packages built:</summary>
  <ul>
    <li>linuxKernel.packages.linux_6_10.msi-ec</li>
    <li>linuxKernel.packages.linux_6_6.msi-ec</li>
    <li>linuxKernel.packages.linux_hardened.msi-ec (linuxKernel.packages.linux_6_6_hardened.msi-ec)</li>
    <li>linuxKernel.packages.linux_libre.msi-ec</li>
    <li>linuxKernel.packages.linux_lqx.msi-ec</li>
    <li>linuxKernel.packages.linux_xanmod.msi-ec</li>
    <li>linuxKernel.packages.linux_xanmod_latest.msi-ec (linuxKernel.packages.linux_xanmod_stable.msi-ec)</li>
    <li>linuxKernel.packages.linux_zen.msi-ec</li>
  </ul>
</details>

**Comment by m1dugh** [comment]
Should the 2 build failures be fixed before merging or not @NotAShelf ?

**Comment by NotAShelf** [comment]
Yes, the build fails on both mainline Linux and Linux-Libre. Neither of those are packages we can skip.

**Comment by m1dugh** [comment]
Here is the result after fix of `nixpkgs-review pr 343038` on `x86_64-linux`

10 packages marked as broken and skipped:
linuxKernel.packages.linux_4_19.msi-ec linuxKernel.packages.linux_4_19_hardened.msi-ec linuxKernel.packages.linux_5_10.msi-ec linuxKernel.packages.linux_5_10_hardened.msi-ec linuxKernel.packages.linux_5_15.msi-ec linuxKernel.packages.linux_5_15_hardened.msi-ec linuxKernel.packages.linux_5_4.msi-ec linuxKernel.packages.linux_5_4_hardened.msi-ec linuxKernel.packages.linux_6_1.msi-ec linuxKernel.packages.linux_6_1_hardened.msi-ec

10 packages built:
linuxKernel.packages.linux_6_10.msi-ec linuxKernel.packages.linux_6_11.msi-ec linuxKernel.packages.linux_6_6.msi-ec linuxKernel.packages.linux_hardened.msi-ec linuxKernel.packages.linux_latest_libre.msi-ec linuxKernel.packages.linux_libre.msi-ec linuxKernel.packages.linux_lqx.msi-ec linuxKernel.packages.linux_xanmod.msi-ec linuxKernel.packages.linux_xanmod_latest.msi-ec linuxKernel.packages.linux_zen.msi-ec

File: maintainers/maintainer-list.nix
Comment (GetPsyched) [inline maintainers/maintainer-list.nix:12509]: You need to set your name here. See CI breakage.

**Comment by m1dugh** [comment]
Any idea of when it can be merged ?

File: pkgs/os-specific/linux/msi-ec/default.nix
Comment (tomodachi94) [reply pkgs/os-specific/linux/msi-ec/default.nix:17]: I'm hesitant to use a fork, especially since the upstream appears to be maintained and [the upstream PR in question which originates from this fork](https://github.com/BeardOverflow/msi-ec/pull/162) has some unresolved change requests from the maintainer.

File: pkgs/os-specific/linux/msi-ec/default.nix
Comment (tomodachi94) [inline pkgs/os-specific/linux/msi-ec/default.nix:21]:
```suggestion
  postPatch = ''
```
Overriding entire phases like this is discouraged ([docs](https://github.com/NixOS/nixpkgs/blob/master/doc/stdenv/stdenv.chapter.md#phases-sec-stdenv-phases)).

File: pkgs/os-specific/linux/msi-ec/patches/Makefile
Comment (tomodachi94) [inline pkgs/os-specific/linux/msi-ec/patches/Makefile:10]:
```suggestion
KERNELDIR ?= /lib/modules/$(KERNEL)/build/
ccflags-y := -std=gnu11 -Wno-declaration-after-statement
```

These are the [flags used by upstream](https://github.com/m1dugh/msi-ec/blob/e261feba4ae92d1626ed5a9c1ce89883b303524c/Makefile#L5). We should use them if possible.

Docs about this: https://www.kernel.org/doc/Documentation/kbuild/makefiles.txt

File: pkgs/os-specific/linux/msi-ec/default.nix
Comment (tomodachi94) [inline pkgs/os-specific/linux/msi-ec/default.nix:34]:
```suggestion
  makeFlags = kernel.makeFlags ++ [
    "KERNELDIR=${kernel.dev}/lib/modules/${kernel.modDirVersion}/build"
    "INSTALL_MOD_PATH=$(out)"
    "ccflags-y+=-Wno-error=incompatible-pointer-types"
  ];
```

Setting NIX_CFLAGS_COMPILE is [usually a last resort](https://discourse.nixos.org/t/env-nix-cflags-compile-vs-cxxflags/39192/8). The Linux kernel Makefile supports ccflags-y, so we should use that instead.

File: pkgs/os-specific/linux/msi-ec/default.nix
Comment (tomodachi94) [inline pkgs/os-specific/linux/msi-ec/default.nix:38]: Is this necessary? At the top of the parched Makefile, we define the `default` phony rule.

**Review by tomodachi94 — CHANGES_REQUESTED** [review]

**Comment by m1dugh** [comment]
@tomodachi94 Should be all good

File: pkgs/os-specific/linux/msi-ec/default.nix
Comment (Tungsten842) [inline pkgs/os-specific/linux/msi-ec/default.nix:30]:
```suggestion

```
This should be unneeded since:
https://github.com/BeardOverflow/msi-ec/commit/a42a91536852c5508d6dd62b2172c55ee9d7edf8

**Review by Tungsten842 — CHANGES_REQUESTED** [review]

**Comment by m1dugh** [comment]
@tomodachi94 The requested changes cannot be resolved due to force-push, if it is good for you can you accept the changes ? :sweat_smile:

**Review by tomodachi94 — APPROVED** [review]
Lgtm, thank you! The initial PR was really good for just your first contribution 🙂

**Review by NotAShelf — APPROVED** [review]

File: pkgs/os-specific/linux/msi-ec/default.nix
Comment (tomodachi94) [inline pkgs/os-specific/linux/msi-ec/default.nix:48]:
```suggestion
    license = lib.licenses.gpl2Plus;
```
See https://github.com/BeardOverflow/msi-ec/blob/main/msi-ec.c#L1. `lib.licenses.gpl2' is deprecated.

**Review by tomodachi94 — CHANGES_REQUESTED** [review]
Something small that I missed in the original review

File: pkgs/os-specific/linux/msi-ec/default.nix
Comment (GetPsyched) [inline pkgs/os-specific/linux/msi-ec/default.nix:6]: I could be wrong here, but can we not simply import `git` instead of `pkgs` and replace `pkgs.git` -> `git` later in the derivation?

**Comment by m1dugh** [comment]
It should be good

**Review by tomodachi94 — APPROVED** [review]
Lgtm, thank you ❤️

**Review by Tungsten842 — APPROVED** [review]

**Review by GetPsyched — APPROVED** [review]

**Pull request merged**
